src/main/java/seedu/taskell/logic/commands/UndoCommand.java

```diff
@@ -1,92 +1,243 @@
  */
 public class UndoCommand extends Command {
     private static final Logger logger = LogsCenter.getLogger(UndoCommand.class.getName());
     
     public static final String COMMAND_WORD = "undo";
-    public static final String MESSAGE_USAGE = COMMAND_WORD + ": Undo most recent command.\n"
-            + "Example: " + COMMAND_WORD;
+    public static final String MESSAGE_USAGE = COMMAND_WORD + ": Undo a previously executed command.\n"
+            + "Parameters: INDEX (must be a positive integer)\n"
+            + "Example: " + COMMAND_WORD + " 3";
     
-    public static final String MESSAGE_DELETE_TASK_SUCCESS = "Deleted Task: %1$s";
-    public static final String MESSAGE_ADD_TASK_SUCCESS = "Task added back: %1$s";
-    public static final String MESSAGE_DUPLICATE_TASK = "This task already exists in the task manager";
-    public static final String MESSAGE_NO_TASK_TO_UNDO = "No add or delete commands available to undo.";
-
-    private static ArrayList<String> commandHistory;
+    private static final String MESSAGE_DELETE_TASK_SUCCESS = "Deleted Task: %1$s";
+    private static final String MESSAGE_ADD_TASK_SUCCESS = "Task added back: %1$s";
+    private static final String MESSAGE_EDIT_TASK_SUCCESS = "Task edited back to old version: %1$s";
+    
+    private static final String MESSAGE_DUPLICATE_TASK = "This task already exists in the task manager";
+    private static final String MESSAGE_NO_TASK_TO_UNDO = "No add or delete commands available to undo.";
+    private static final String MESSAGE_COMMAND_HISTORY_EMPTY = "No command history available for undo.";
+    private static final String MESSAGE_INVALID_INDEX = "Index is invalid";
+    
+    private static ArrayList<CommandHistory> commandHistoryList;
+    
+    private int index;
+    private CommandHistory commandHistory;
     
-    private static String mostRecentCommand;
-    private static Task mostRecentAddedTask;
-    private static Task mostRecentDeletedTask;
+    public UndoCommand(int index) {
+        logger.info("Creating UndoCommand with index: " + index);
+        this.index = index;
+    }
     
-    public UndoCommand() {
+    public static ArrayList<String> getListOfCommandHistoryText() {
+        assert commandHistoryList != null;
+        assert !commandHistoryList.isEmpty();
+        
+        ArrayList<String> list = new ArrayList<>();
+        for (CommandHistory history: commandHistoryList) {
+            list.add(history.getCommandText());
+        }
         
+        return list;
     }
     
     @Override
     public CommandResult execute() {
-        if (mostRecentCommand==null) {
-            return new CommandResult(String.format(MESSAGE_NO_TASK_TO_UNDO));
+        
+        if (commandHistoryList.isEmpty()) {
+            return new CommandResult(String.format(MESSAGE_COMMAND_HISTORY_EMPTY));
+        }
+        else if (index > commandHistoryList.size()) {
```

Comment:
else if should comes after '{'
![image](https://cloud.githubusercontent.com/assets/13817632/19644061/4cab02ea-9a20-11e6-897c-d96a3a7f5acb.png)